Define API compatibility story

tigrannajaryan opened the pull request:
We need to define the API compatibility story, particularly:

- How OpenTelemetry API evolves over time, what changes are allowed and how they should be made, how breaking and non-breaking changes are handled.
- How we ensure third party libraries that are instrumented using old OpenTelemetry versions remain usable after the API changes.
- How we ensure interoperability of application code and of multiple third party libraries that are instrumented using different OpenTelemetry versions.

Go toy example: https://github.com/tigrannajaryan/otelapi-compat
Java toy example:  https://github.com/tigrannajaryan/otelapi-java-compat

Implements https://github.com/open-telemetry/opentelemetry-specification/issues/109

tigrannajaryan commented on specification/compatibility.md:
```diff
@@ -0,0 +1,38 @@
+
+New functionality in minor version increases may be added in the form of new classes, methods or functions in a way that does not affect existing functionality.
+
+The first OpenTelemetry API release will be version 1.0.0.
```
Question: is this what we want to do or we want to start with v0?

yurishkuro replied:
I suggest we start with 0.x and gather feedback first. We may not get v1 done right right away.

iredelmeier replied:
Big +1 re: starting with 0.x.0, where x may be >= 1 in September. I think it's fine - maybe even better? - if we start using semver sooner rather than later so that we can suss out versioning questions sooner rather than later.

tigrannajaryan replied:
Is the expectation here that we simply want to start from version number 0 and thus everything else written in this document stays the same, we simply map v1 to v0, v2 to v1, etc?

Or the expectation is that version 0 has a different behaviour from compatibility perspective compared to version 1 and newer?

@yurishkuro @iredelmeier @reyang @austinlparker @carlosalberto can you elaborate on what your expectation is? If v0 must be in principle different from v1 and newer then what exactly do you expect to be different?

yurishkuro replied:
The expectations of v. 0.x.x are different from 1.x+. In 0.x breaking changes are expected between minor versions.

tigrannajaryan replied:
Fixed description of 0.x case.

tigrannajaryan commented on specification/compatibility.md:
```diff
@@ -0,0 +1,38 @@
+Any application or third party library that uses OpenTelemetry API of a particular version can safely be packaged with OpenTelemetry SDK implementation provided that these 2 conditions are met:
+
+1. The full version number of the SDK is not lower than the full version number of API used by application or third party library.
+2. The major version number of the SDK is the same as major version number of API or major version number of the SDK is higher by 1.
```
Question: do we want to provide stronger guarantees across more than one major version number? Or perhaps we don't want to provide compatibility guarantees across major versions at all? 
I felt that one major version apart is a good approach.

yurishkuro replied:
Do we want to tightly couple API version with SDK version? We may need to do major rev of SDK because some of its SPI interfaces change, while the API remains unchanged.

tigrannajaryan replied:
I think we should not tightly couple API and SDK versions. The rationale is that you cannot expect all third party libraries that a developer may use in an application to all have been instrumented against the same OpenTelemetry API. Third-party developers move at their own pace, you usually move at your own pace on your own app. There has to be a way for mixed-version usage. At the minimum code linked to APIs across minor versions should be interoperable, but I think that is not good enough and 1 major version apart is the minimum guarantee that is necessary for real-life usage.

yurishkuro replied:
+1

yurishkuro replied:
To clarify, the current wording is still "tight coupling" to me, as it implies correlation between API and SDK versions. To take example of OpenTracing: Python API is v2, Jaeger SDK is v4. Granted that in OpenTelemetry the API and SDK are more closely related, but there are still potential major versions of SDK (e.g. changing from JS to Typescript as the original source) that do not affect major rev of the API.

This is separate from the point that a version of SDK must support some V and V-1 of the API.

songy23 commented on specification/compatibility.md:
```diff
@@ -0,0 +1,38 @@
+
+## Semantic Versioning of API
+
+OpenTelemetry API follows [semantic versioning convention](https://semver.org/spec/v2.0.0.html). Patch number in API version is incremented for bug fixes, minor version number is incremented when new, backwards compatible functionality is added, major version number is incremented when breaking, non-backwards compatible changes are done.
```
```suggestion
OpenTelemetry API follows [semantic versioning convention](https://semver.org/spec/v2.0.0.html). 
- Patch number in API version is incremented for bug fixes.
- Minor version number is incremented when new, backwards compatible functionality is added.
- Major version number is incremented when breaking, non-backwards compatible changes are done.
```

yurishkuro replied:
I feel this explanation is obvious, but not helpful, since it leaves out the hard part: the definition of backwards compatible change. Is adding a method to an interface backwards compatible? It breaks implementors of the interface, which can be both vendors and end users.

tigrannajaryan replied:
"backwards compatible" in the sense that the change in a dependency does not break the dependent code. I think in your example it is clear that adding a method to an interface is a breaking change. Do you think it is worth adding more detail and particularly to define that "backwards compatible" means "not breaking" the user code?

yurishkuro replied:
I think defining what we mean by "breaking" is more important than restating what semver means (we're not saying anything new there). E.g. in OpenTracing  we considered adding methods non-breaking because it was not breaking the _end user / instrumentation code_, even though it was a breaking change for vendors implementing OT API.

tigrannajaryan replied:
I understand now. Everything in this document is written from the perspective of end users of the API. Your raised a good point, let me think about vendor side of things and get back.

tigrannajaryan replied:
I added a separate issue to define compatibility rules for vendor APIs: https://github.com/open-telemetry/opentelemetry-specification/issues/128

For the sake of making progress I suggest that we work on it separately, after the compatibility story for end-user instrumentation API is defined and this PR is merged.

iredelmeier replied:
👍

tigrannajaryan replied:
I also added a separate section about 0.x. Resolving this.

tigrannajaryan commented on specification/compatibility.md:
```diff
@@ -0,0 +1,38 @@
+
+## API Namespacing
+
+The OpenTelemetry API version 1.x.x is defined in io.opentelemetry namespace. The second major version of the API will be in io.opentelemetry.v2 namespace.
```
Are we OK with namespacing?

yurishkuro replied:
I don't think it works in all languages. Also, what is the implication of using two versions in one app - are they supposed to interop (causes circular dep)? If that's not the goal then why else is package namespacing needed?

tigrannajaryan replied:
>I don't think it works in all languages.

Is there a language where at least _some_ form of namespacing is not possible? Worst case you can use prefixes or suffixes for all function names, e.g. CreateSpan, CreateSpanV2, etc.

>Also, what is the implication of using two versions in one app - are they supposed to interop (causes circular dep)?

Yes, I suggest that they must interop and the SDK must guarantee that by providing implementations of previous API versions. See example in this doc.

tigrannajaryan replied:
>causes circular dep

Circular dep between which components?

carlosalberto replied:
> CreateSpanV2

I personally think this is bad in many ways ;)

I's stick with some form of namespacing, if truly needed.

tigrannajaryan replied:
>    CreateSpanV2
>I personally think this is bad in many ways ;)

I am not advocating to use suffixes in function names as a superior way. I mentioned it as a last resort approach when the language does not provide formal constructs for namespacing, which is rare, but happens (e.g. C) but probably doesn't apply to any of the languages that OpenTelemetry will support (I didn't do exhaustive check).

iredelmeier replied:
(As above, I believe this conversation could be pulled out so it can be discussed outside of the PR)

I think it's confusing to specify the namespace here as it will need to be handled differently across languages.

An idea I've been toying with is something where, while we're still in the earlier phases, we provide versioned packages that **wrap** newer versions as they get released - i.e., `opentelemetry/v0.1.0` would always give you the same interface, but would internally wrap `opentelemetry/v0.2.0`... which would in turn wrap `opentelemetry/v0.3.0`.
* + This would allow users (including integrations!!!) to pin to versions, while...
* + ...allowing us to iterate faster without breaking users to the same degree
* - it limits the sorts of changes we can make

(To clarify: this is more of an idea for SDK versioning, but I think if we go this route, it could enable us to do something similar-ish for API versions?)

yurishkuro replied:
+1 on separate issue and a TODO here.

@iredelmeier the benefit of namespacing is that it makes fully qualified API types clearly separate, which actually enables the wrappers for up/down versions, which otherwise are difficult since the same class names cannot have two meanings in a single binary. However, @tigrannajaryan, the downside of namespacing is that it makes migrations unnecessarily more complicated, because all instrumentation code needs to be refactored to point to a new namespace, even though that code may not be using any APIs that were actually changed/broken between V and V+1.

tigrannajaryan replied:
Sounds good, added a separate issue here: https://github.com/open-telemetry/opentelemetry-specification/issues/129

iredelmeier replied:
@yurishkuro I agree re: the value of namespacing. My concern was specifically about referring to it in the spec in a language-specific way, i.e., `io.opentelemetry`.

tigrannajaryan replied:
Resolving, this a separate issue.

yurishkuro commented on specification/compatibility.md:
```diff
@@ -0,0 +1,38 @@
+
+The above is specified using Java's convention of reverse domain name notation. Other languages will use their equivalent convention for namespaces, e.g. for Go the package name for version 1.x.x of the API will be go.opencensus.io, for version 2.x.x of the API it will be go.opencensus.io/v2.
+
+## Interoperability of Mixed-version Code
```
I am not sure how this would work. This might work for libs that are used at the leaves of the call graph, but what if we have server middleware storing v2 span in context, and http client lib middleware that expects v1 span?

tigrannajaryan replied:
Can you elaborate on this? I am not sure I fully understand the example that you brought.

yurishkuro replied:
so in this picture:
![image](https://user-images.githubusercontent.com/3523016/59400404-4e840900-8d65-11e9-98c5-ffa829b3e27b.png)

* Handler instrumentation is v2 and stores `io.opentelemetry.v2.Span` in the Context
* Client instrumentation is v2 and tries to read `io.opentelemetry.v1.Span` from the Context, which is not going to work.

tigrannajaryan replied:
Thanks a lot for the picture, it is clear now.

I do not see this as a problem that is induced by the decision to support multiple API versions in parallel in OpenTelemetry and let me explain why I think so.

Based on your description I am going to assume that Context is a container that has a pair of GetSpan/SetSpan methods that allow to store and retrieve arbitrary untyped value of Span (an Object in Java's parlance or interface{} in Go's).

In this situation I believe the problem is a result of incorrect assumptions of Client which reads an untyped data type from Context and treats it as if it has a certain type when in fact it can be of any type.

If the Context is a container of untyped data types where I can store 2 unrelated data types (io.opentelemetry.v2.Span and io.opentelemetry.v1.Span) then users of this container (namely Handler and Client) cannot expect that the untyped container will magically solve their interoperability problem.

If Handler and Client want to interchange data the responsibility is on them to define a data exchange contract. I fail to see how it can be the responsibility of generic Context class which is merely a carrier of untyped data.

Now, if we assume the opposite, i.e. that Context is a strongly typed container and supports GetSpan/SetSpan strongly typed methods then this problem cannot exist because Context itself must be versioned, so we will have io.opentelemetry.v1.Context and io.opentelemetry.v2.Context, each having a pair of GetSpan/SetSpan methods that work only with corresponding data type. In that case the SDK can ensure interoperability of v1.Context and v2.Context types by converting v1.Span into v2.Span internally (but I am not sure it should try to, that's a separate question worth thinking about).

I may be wrong, happy to hear your thoughts.

tigrannajaryan replied:
Here I created an example of what OpenTelemetry API V2 implementation can look like, so that it provides a shim for API V1 and how exactly it can be used in the scenario that you described: with a middleware that uses API V1 and an app client that uses API V2.

It shows how the shim ensures the interoperability by mapping the API V1 Span methods to API V2 Span methods as needed.

Check this out: https://github.com/tigrannajaryan/otelapi-compat

tigrannajaryan replied:
@yurishkuro I wonder if you had a chance to have a look at the sample code I posted which shows how the scenario you described would work with the compatibility approach that I suggest in this PR.

yurishkuro replied:
@tigrannajaryan in your example, try to add a v1 client-span, e.g. from instrumentation for an imaginary database call. Or just try to implement `opentelemetryv1.FromContext`, which is missing.

tigrannajaryan replied:
@yurishkuro Here is the implementation of `opentelemetryv1.FromContext`:
https://github.com/tigrannajaryan/otelapi-compat/blob/849e405dcddf73ca2f1cfdc429d03dde504dd9cc/opentelemetryv1/tracer.go#L37

Imaginary database call that creates v1 client-span (initiated from v2-instrumented call): https://github.com/tigrannajaryan/otelapi-compat/blob/849e405dcddf73ca2f1cfdc429d03dde504dd9cc/myapp/main.go#L35

yurishkuro replied:
How would this work with go modules? The v2 API needs to include a v1 package whose implementation is different from the real (old) v1.

tigrannajaryan replied:
(OK, now I can reply to the comment, strange, so copying my reply here).

There are probably different ways to achieve this in Go. One approach is to have an additional indirection level from API to SDK. This follows from the the requirement to be able to swap out SDK implementations laid out here: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/library-guidelines.md

If you have an indirection layer in API v1 which allows registering implementation at runtime, then the code in SDK v2 will register its own implementation of API v1 at startup.

I am not sure if there is a way to perform this swapping in Go at compile time, at least I cannot see it yet.

For Java this is easier to do since you can just provide a different jar which implements the exact same API in a different, so there is no need to have runtime swapping, it can happen at compile time.

Just to reiterate, this is not a requirement that comes from this PR, it is something that has been agreed to be part of the spec in the past.

tigrannajaryan replied:
I can modify my example source code to illustrate what I mean if the description is not clear.

tigrannajaryan replied:
@yurishkuro I modified my toy example at https://github.com/tigrannajaryan/otelapi-compat/ to demonstrate how API and SDK can be decoupled and how v2 SDK can implement both v1 and v2 APIs and make it all work on mixed-code app without touching third-party library code.

tigrannajaryan commented on specification/compatibility.md:
```diff
@@ -0,0 +1,38 @@
+
+## Compatibility of SDK Implementations
+
+Any application or third party library that uses OpenTelemetry API of a particular version can safely be packaged with OpenTelemetry SDK implementation provided that these 2 conditions are met:
```
Fixed, thanks.

tigrannajaryan commented:
Everyone, I would like to have a discussion about API compatibility in the next week's specification SIG meeting. I believe it is important to define this early, otherwise compatibility issues will bite us back in the future. Please join the meeting on Tue, June 18, 8am PT.

Agenda and Zoom link: https://docs.google.com/document/d/1-bCYkN-DWJq4jw1ybaDZYYmx-WAe6HnwfWbkm8d57v8/edit#heading=h.v6lhw87olr5l

owais commented on specification/compatibility.md:
```diff
@@ -0,0 +1,48 @@
+1. The full version number of the SDK is not lower than the full version number of API used by application or third party library.
+2. The major version number of the SDK is the same as major version number of API or major version number of the SDK is higher by 1.
+
+For example, an application that uses API 1.5.10 can be packaged with SDK 1.5.12 or SDK 2.0.0, but it cannot be packaged with SDK 1.0.0 (too old) or with SDK 3.0.0 (too new).
```
Why would 3.0 be too new for 1.5 but 2.0 would be fine? They could both break things in equally bad ways right?

tigrannajaryan replied:
The more old versions you decide to support the more maintenance work it requires. I believe supporting one previous major version is a good balance, it gives people who depend on OpenTelemtry enough time to upgrade, yet allows us to have manageable amount of old / deprecated functionality that we need to maintain.

carlosalberto replied:
Aside the approach we end up taking, I do agree on @tigrannajaryan approach: one previous major version to support.

tigrannajaryan commented:
Reviewers, I think I addressed all comments. Please have a look again. If there are no other questions I'd like to merge this and address any outstanding issues separately (I filed a couple as a result of discussions we had on this PR).

yurishkuro commented on specification/compatibility.md:
```diff
@@ -0,0 +1,48 @@
+
+New functionality in minor version increases may be added in the form of new classes, methods or functions in a way that does not affect existing functionality.
+
+## Compatibility of SDK Implementations
```
We should be separating API and SDK specs.

tigrannajaryan commented:
>How would this work with go modules? The v2 API needs to include a v1 package whose implementation is different from the real (old) v1.

@yurishkuro  For some reason I cannot reply to this comment directly, so replying here.

There are probably different ways to achieve this in Go. One approach is to have an additional indirection level from API to SDK. This follows from the the requirement to be able to swap out SDK implementations laid out here: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/library-guidelines.md

If you have an indirection layer in API v1 which allows registering implementation at runtime, then the code in SDK v2 will register its own implementation of API v1 at startup.

I am not sure if there is a way to perform this swapping in Go at compile time, at least I cannot see it yet.

For Java this is easier to do since you can just provide a different jar which implements the exact same API in a different, so there is no need to have runtime swapping, it can happen at compile time.

Just to reiterate, this is not a requirement that comes from this PR, it is something that has been agreed to be part of the spec in the past.

cstockton commented:
@tigrannajaryan I appreciate the work you are doing here, really, I don't want to detract from the fact you are writing documentation on the current design. Thank you.

That said maybe you can shed some light on why we are defining this at a software integration level. This is going to require massive engineering efforts and is already influencing "plugin" based design in the client libraries like https://github.com/open-telemetry/opentelemetry-go/issues/19. This specification you are writing is essentially going to devolve into an aggregation of `man ld*` best practices, e.g. `foo.so.X`. Consequentially OpenTelemetry will struggle to add new features because ABI is as brittle as it comes across every potential architecture and system then a network transport.

I am trying to follow these issues and conversations but I just can't seem to find what the objections to defining the "Tracer Transport Protocol" (the name is for illustration) whatever that may be. If we can take a small bit of time to do this, it lowers the work across the board and sets a solid foundation to build on. Most importantly it doesn't stop OpenTelemetry at that point to writing monolithic SDK's which integrate many vendors and platforms **exactly** as they are today. But it does provide a standard specification for **users** that disagree with such a design from having to integrate with them.

What are you thoughts on:
1) pause momentarily
2) define TTP
3) implementations of TTP, use them to inform changes in 2
4) unpause - Resume exact plans for SDK's, which push traces over TTP

If <Insert trace vendor> wants to support the SDK's they can, just as they do under the current design. However they now have the option to add a flag on their existing server endpoints to accept those trace formats `--ttp-address`, allowing them to leverage their existing libraries in the language they have the most resources in and avoiding the potential bureaucracy of "getting permission" from OpenTelemetry. I can write a client to talk to said vendor and cut out the middleman if I want to. OpenTelemetry may provide enough value for me to choose to use it for a given language, but I have a choice. The model just scales better and will result in best-in-class solutions to each problem. Maybe the most reliable kafka ingest server lands at <$USER>/ttp2kafka and is written in rust. A highly reliable / configurable tpp2es ships to Elasticsearch with TLS support written in nodejs. Cool, they are just servers and we only need a single reliable client in each language.

Small note: I would be more than happy to contribute to helping define the specification if I knew it wouldn't be dismissed due to [we can always do that later](https://github.com/open-telemetry/opentelemetry-specification/issues/20#issuecomment-504773703).

tigrannajaryan commented:
Reviewers, please comment.

I'd like to move forward. If this is proposal is acceptable in principle let's merge it and we can improve it in additional PRs.

If it is not acceptable let's reject it, I am OK with that too.

I strongly believe that we need to have a well-defined compatibility story (not necessarily exactly the story suggested in this PR). The story needs to be publicly available. People who evaluate OpenTelemetry for use in their applications or libraries will look into our compatibility guarantees as part of their evaluation. Not having clearly articulated compatibility specification is not helping us.

tigrannajaryan commented:
@yurishkuro as discussed I have put together the Java version of the API/SDK version compatibility demo: https://github.com/tigrannajaryan/otelapi-java-compat

It is less elaborate than the Go version (e.g. Context propagation is not implemented) but I believe the idea is clear anyway.

Please let me know if you have any other concerns.

SergeyKanzhelev commented:
@tigrannajaryan can you move this forward? Or we can abandon this PR for now

tigrannajaryan commented:
@SergeyKanzhelev let me think a bit about how this will work for our recently agreed version labeling and I will post an update to this PR or ask for postponing it.

SergeyKanzhelev commented:
@tigrannajaryan I'll close this PR for now than. It shouldn't be a problem to re-open it.

tigrannajaryan commented:
ok, thanks @SergeyKanzhelev
